packages/mds-db/events.ts

```diff
@@ -304,6 +304,92 @@ export async function readEventsWithTelemetry({
   }))
 }
 
+// TODO: remove
+// heinous copypasta specifically to provide the VIN with every event, used ONLY by Native.
+// this is to be excised when we dump Native in the garbage in the 1.0 timeframe.
+export async function readEventsWithTelemetryAndVehicleId({
+  device_id,
+  provider_id,
+  start_time,
+  end_time,
+  order_by = 'id',
+  last_id = 0,
+  limit = 1000
+}: Partial<{
+  device_id: UUID
+  provider_id: UUID
+  start_time: Timestamp
+  end_time: Timestamp
+  order_by: string
+  last_id: number
+  limit: number
+}>): Promise<Recorded<VehicleEvent & Pick<Device, 'vehicle_id'>>[]> {
+  const client = await getReadOnlyClient()
+  const vals = new SqlVals()
+  const exec = SqlExecuter(client)
+
+  const conditions: string[] = last_id ? [`id > ${vals.add(last_id)}`] : []
+
+  if (provider_id) {
+    if (!isUUID(provider_id)) {
+      throw new Error(`invalid provider_id ${provider_id}`)
+    } else {
+      conditions.push(`provider_id = ${vals.add(provider_id)}`)
+    }
+  }
+
+  if (device_id) {
+    if (!isUUID(device_id)) {
+      throw new Error(`invalid device_id ${device_id}`)
+    } else {
+      conditions.push(`device_id = ${vals.add(device_id)}`)
+    }
+  }
+
+  if (start_time !== undefined) {
+    if (!isTimestamp(start_time)) {
+      throw new Error(`invalid start_time ${start_time}`)
+    } else {
+      conditions.push(`timestamp >= ${vals.add(start_time)}`)
+    }
+  }
+
+  if (end_time !== undefined) {
+    if (!isTimestamp(end_time)) {
+      throw new Error(`invalid end_time ${end_time}`)
+    } else {
+      conditions.push(`timestamp <= ${vals.add(end_time)}`)
+    }
+  }
+
+  // we can only select based on event criteria
+  const where = conditions.length ? ` WHERE ${conditions.join(' AND ')}` : ''
+
+  const { rows } = await exec(
+    `SELECT E.*, D.vehicle_id, T.lat, T.lng, T.speed, T.heading, T.accuracy, T.altitude, T.charge, T.timestamp AS telemetry_timestamp 
+      FROM (SELECT * FROM ${schema.TABLE.events}${where} ORDER BY ${order_by} LIMIT ${vals.add(limit)}) AS E 
+    LEFT JOIN ${schema.TABLE.devices} D ON E.device_id = D.device_id
+    LEFT JOIN ${schema.TABLE.telemetry} T ON E.device_id = T.device_id 
+      AND CASE WHEN E.telemetry_timestamp IS NULL THEN E.timestamp ELSE E.telemetry_timestamp END = T.timestamp
```

Comment:
For future reference, `COALESCE(E.telemetry_timestamp, E.timestamp)` which returns the first non-NULL value can be used as a shorthand version of this particular `CASE` pattern,